refactor(acceptance): language-agnostic fix prompts + hardening log hygiene

**Severity**: 🟡 medium (Bug 6) + 🟢 low (Bug 8). Bundled because both are in the acceptance subsystem and Bug 8 is too small for its own PR.

**Linked**: [`docs/findings/2026-05-05-context-curator-v0-dogfood-findings.md`](docs/findings/2026-05-05-context-curator-v0-dogfood-findings.md) → Bug 6, Bug 8
**Patch plan**: [`docs/findings/2026-05-05-acceptance-fix-prompt-language-agnostic.md`](docs/findings/2026-05-05-acceptance-fix-prompt-language-agnostic.md)

## Bug 6 — Acceptance fix prompt is 89 KB / 1891 lines

The largest observed prompt during dogfood:
- Lines 19–543: full test runner output, **including all 36 passing-test lines** (`(pass) AC-N: …`)
- Lines 548–1868: complete 1320-line test file body embedded in a hardcoded ` ```typescript ` fence
- Single turn took 1240 s ($2.71); session forced mid-task `Compacting…`
- Agent re-read the file via tools anyway: *"I need the actual code, not summaries. Let me read the key files directly."*

### Two separate problems in one builder

1. **Test output is raw** — should be filtered to failures only, language-agnostic
2. **Test file body is embedded** — agent has the path + Read access; embedding is double work
3. **Code fence is hardcoded `typescript`** — wrong for `_test.go`, `test_foo.py`, etc.

### Good news — the SSOT we need already exists

`src/test-runners/parser.ts`:
- `parseTestOutput(output)` — auto-detects bun/jest/vitest/pytest/go/unknown, returns structured `TestSummary`
- `formatFailureSummary(failures, maxChars=2000)` — pre-built formatter with truncation
- `detectFramework(output)` — language-agnostic detection
- `buildTestFrameworkHint(testCommand)` — for prompt framing

The fix is plumbing, not new logic.

### Proposed fix

New `src/prompts/builders/acceptance-builder-helpers.ts`:
- `formatTestOutputForFix(rawOutput)` — 4-tier fallback: structured → tail → environmental → summary
- `fenceLangFor(testPath)` — derive code fence from file extension

Then in `src/prompts/builders/acceptance-builder.ts`:
- `buildTestFixPrompt`, `buildSourceFixPrompt`, `buildDiagnosisPromptTemplate` — pipe output through `formatTestOutputForFix`; derive fence from path; replace embedded body with path-only reference + Read instruction

In `src/operations/acceptance-fix.ts`:
- Drop `testFileContent` from `AcceptanceFixSourceInput` / `AcceptanceFixTestInput`; add `testCommand`

### Expected impact

89 KB prompt → ~5–10 KB. No more mid-task `Compacting…`. Single-turn cost drops from $2.71 toward ~$0.30.

| Input | Today | After |
|:---|:---|:---|
| 7 TS failures (bun runner) | All 36 pass lines + 1320-line test file + raw output | 7 structured failures (~3 KB) + path-only reference |
| 4 Go test failures (`_test.go`) | hardcoded ` ```typescript ` fence; raw `--- FAIL:` output | Go parser extracts `--- FAIL: TestName`; fence becomes `go` |
| 12 Python failures (pytest) | TS fence, raw output | pytest parser, `python` fence |
| Unknown runner (rspec/phpunit/jasmine) | Whole output dumped | `parseCommonOutput` regex fallback; tail-N lines |

---

## Bug 8 — `Hardening pass complete` log emitted twice

`runs/2026-05-04T15-10-16.jsonl:185-186`:
```
{"timestamp":"2026-05-04T15:48:08.596Z","stage":"acceptance","message":"Hardening pass complete","data":{"storyId":"US-001","promoted":23,"discarded":5}}
{"timestamp":"2026-05-04T15:48:08.596Z","stage":"acceptance","message":"Hardening pass complete","data":{"storyId":"US-001","promoted":23,"discarded":5}}
```

Identical timestamps, identical data — same event logged twice from two sites:
1. `src/acceptance/hardening.ts:210-214` (inside the function)
2. `src/pipeline/stages/acceptance.ts:267-271` (at the stage layer after calling the function)

Bonus issue: both lines attribute to a single `storyId` even though the operation iterates over **all stories with `suggestedCriteria`** — `storiesWithSuggested[0].id` vs `ctx.story.id` can diverge.

### Proposed fix

1. Drop the redundant stage-level emit in `acceptance.ts:267-277` (function already logs)
2. Fix attribution in `hardening.ts:210-214` to use `storyIds: [...]` and `storiesProcessed: N`
3. Same fix for the corresponding `Hardening pass failed` warn-emit pair

## Files affected

- `src/prompts/builders/acceptance-builder-helpers.ts` *(new)*
- `src/prompts/builders/acceptance-builder.ts`
- `src/operations/acceptance-fix.ts`
- `src/acceptance/fix-executor.ts` (caller — pass `testCommand`, stop reading the test file)
- `src/acceptance/hardening.ts` (Bug 8 part 1)
- `src/pipeline/stages/acceptance.ts` (Bug 8 part 2)

Full diff plan, test plan, risk assessment, commit plan (9 commits across both bugs), follow-ups (F1: same fix for `RectifierPromptBuilder`; F2: extend per-runner parsers for multi-line diff blocks; F3: SSOT `fenceLangFor`; F4: audit other prompt builders) in [`docs/findings/2026-05-05-acceptance-fix-prompt-language-agnostic.md`](docs/findings/2026-05-05-acceptance-fix-prompt-language-agnostic.md).

--- a/src/acceptance/hardening.ts
+++ b/src/acceptance/hardening.ts
@@ -58,8 +58,8 @@ export async function runHardeningPass(ctx: HardeningContext): Promise<Hardening
   if (storiesWithSuggested.length === 0) return result;
 
   logger?.info("acceptance", "Starting hardening pass", {
-    storyId: storiesWithSuggested[0].id,
-    storiesWithSuggested: storiesWithSuggested.length,
+    storyIds: storiesWithSuggested.map((s) => s.id),
+    storiesProcessed: storiesWithSuggested.length,
     totalSuggestedACs: storiesWithSuggested.reduce((n, s) => n + (s.suggestedCriteria?.length ?? 0), 0),
   });
 
@@ -134,7 +134,8 @@ export async function runHardeningPass(ctx: HardeningContext): Promise<Hardening
         language,
       );
       logger?.warn("acceptance", "Hardening generate op returned no test code — using skeleton", {
-        storyId: storiesWithSuggested[0].id,
+        storyIds: storiesWithSuggested.map((s) => s.id),
+        storiesProcessed: storiesWithSuggested.length,
       });
     }
     await _hardeningDeps.writeFile(suggestedTestPath, testCode);

--- a/src/acceptance/hardening.ts
+++ b/src/acceptance/hardening.ts
@@ -208,13 +208,15 @@ export async function runHardeningPass(ctx: HardeningContext): Promise<Hardening
     }
 
     logger?.info("acceptance", "Hardening pass complete", {
-      storyId: storiesWithSuggested[0].id,
+      storyIds: storiesWithSuggested.map((s) => s.id),
+      storiesProcessed: storiesWithSuggested.length,
       promoted: result.promoted.length,
       discarded: result.discarded.length,
     });
   } catch (err) {
     logger?.warn("acceptance", "Hardening pass failed (non-blocking)", {
-      storyId: storiesWithSuggested[0].id,
+      storyIds: storiesWithSuggested.map((s) => s.id),
+      storiesProcessed: storiesWithSuggested.length,
       error: err instanceof Error ? err.message : String(err),
     });
   }

--- a/src/execution/lifecycle/acceptance-fix.ts
+++ b/src/execution/lifecycle/acceptance-fix.ts
@@ -46,6 +46,7 @@ export interface ResolveAcceptanceDiagnosisOptions {
   diagnosisOpts: {
     testOutput: string;
     testFileContent: string;
+    acceptanceTestPath?: string;
     workdir: string;
     storyId?: string;
   };
@@ -117,6 +118,7 @@ export async function resolveAcceptanceDiagnosis(opts: ResolveAcceptanceDiagnosi
   return await _diagnosisDeps.callOp(fixCallCtx(ctx), acceptanceDiagnoseOp, {
     testOutput: diagnosisOpts.testOutput,
     testFileContent: diagnosisOpts.testFileContent,
+    acceptanceTestPath: diagnosisOpts.acceptanceTestPath,
     sourceFiles,
     semanticVerdicts,
   });

--- a/src/execution/lifecycle/acceptance-loop.ts
+++ b/src/execution/lifecycle/acceptance-loop.ts
@@ -210,8 +210,8 @@ export async function runAcceptanceFixCycle(
   prd: PRD,
   initialFailures: { failedACs: string[]; testOutput: string },
   diagnosis: DiagnosisResult,
-  testFileContent: string,
   acceptanceTestPath: string,
+  testCommand?: string,
 ): Promise<FixCycleResult<Finding>> {
   const runtime = ctx.runtime;
   if (!runtime) {
@@ -235,10 +235,10 @@ export async function runAcceptanceFixCycle(
         fixOp: acceptanceFixSourceOp,
         buildInput: (_findings, priorIterations, _ctx) => ({
           testOutput: currentTestOutput,
+          testCommand,
           diagnosisReasoning: diagnosis.reasoning,
           priorIterationsBlock: buildPriorIterationsBlock(priorIterations),
           acceptanceTestPath,
-          testFileContent,
         }),
         maxAttempts: 3,
         coRun: "co-run-sequential",
@@ -250,11 +250,11 @@ export async function runAcceptanceFixCycle(
         fixOp: acceptanceFixTestOp,
         buildInput: (_findings, priorIterations, _ctx) => ({
           testOutput: currentTestOutput,
+          testCommand,
           diagnosisReasoning: diagnosis.reasoning,
           priorIterationsBlock: buildPriorIterationsBlock(priorIterations),
           failedACs: currentFailedACs,
           acceptanceTestPath,
-          testFileContent,
         }),
         maxAttempts: 3,
         coRun: "co-run-sequential",
@@ -419,12 +419,13 @@ export async function runAcceptanceLoop(ctx: AcceptanceLoopContext): Promise<Acc
       );
     }
 
-    // Load test file content for diagnosis
+    // Load test file content for diagnosis (still needed for import parsing in loadSourceFilesForDiagnosis)
     const testEntries = ctx.acceptanceTestPaths
       ? await loadAcceptanceTestContentModule(ctx.acceptanceTestPaths.map((p) => p.testPath))
       : [];
     const testFileContent = testEntries[0]?.content ?? "";
     const acceptanceTestPath = testEntries[0]?.testPath ?? ctx.acceptanceTestPaths?.[0]?.testPath ?? "";
+    const testCommand = ctx.config.quality?.commands?.test;
 
     const strategy = ctx.config.acceptance.fix?.strategy ?? "diagnose-first";
     const diagnosis = await resolveAcceptanceDiagnosis({
@@ -436,6 +437,7 @@ export async function runAcceptanceLoop(ctx: AcceptanceLoopContext): Promise<Acc
       diagnosisOpts: {
         testOutput: failures.testOutput,
         testFileContent,
+        acceptanceTestPath,
         workdir: ctx.workdir,
         storyId: firstStory?.id,
       },
@@ -449,7 +451,7 @@ export async function runAcceptanceLoop(ctx: AcceptanceLoopContext): Promise<Acc
     });
 
     // ── 5. Run acceptance fix cycle ────────────────────────────────────
-    const cycleResult = await runAcceptanceFixCycle(ctx, prd, failures, diagnosis, testFileContent, acceptanceTestPath);
+    const cycleResult = await runAcceptanceFixCycle(ctx, prd, failures, diagnosis, acceptanceTestPath, testCommand);
     // Cost is captured at the dispatch-bus layer (runtime.costAggregator); the local
     // accumulation here is best-effort and may undercount. The authoritative total
     // is reconciled in handleRunCompletion via Math.max(local, aggregator).

--- a/src/operations/acceptance-diagnose.ts
+++ b/src/operations/acceptance-diagnose.ts
@@ -10,6 +10,7 @@ import type { RunOperation } from "./types";
 export interface AcceptanceDiagnoseInput {
   testOutput: string;
   testFileContent: string;
+  acceptanceTestPath?: string;
   sourceFiles: Array<{ path: string; content: string }>;
   semanticVerdicts?: SemanticVerdict[];
 }
@@ -39,6 +40,7 @@ export const acceptanceDiagnoseOp: RunOperation<AcceptanceDiagnoseInput, Accepta
     const prompt = new AcceptancePromptBuilder().buildDiagnosisPrompt({
       testOutput: input.testOutput,
       testFileContent: input.testFileContent,
+      acceptanceTestPath: input.acceptanceTestPath,
       sourceFiles: input.sourceFiles,
       semanticVerdicts: input.semanticVerdicts,
     });

--- a/src/operations/acceptance-fix.ts
+++ b/src/operations/acceptance-fix.ts
@@ -5,19 +5,19 @@ import type { RunOperation } from "./types";
 
 export interface AcceptanceFixSourceInput {
   testOutput: string;
+  testCommand?: string;
   diagnosisReasoning?: string;
   priorIterationsBlock?: string;
   acceptanceTestPath: string;
-  testFileContent?: string;
 }
 
 export interface AcceptanceFixTestInput {
   testOutput: string;
+  testCommand?: string;
   diagnosisReasoning?: string;
   priorIterationsBlock?: string;
   failedACs: string[];
   acceptanceTestPath: string;
-  testFileContent?: string;
 }
 
 export interface AcceptanceFixOutput {
@@ -35,10 +35,10 @@ export const acceptanceFixSourceOp: RunOperation<AcceptanceFixSourceInput, Accep
   build(input, _ctx) {
     const prompt = new AcceptancePromptBuilder().buildSourceFixPrompt({
       testOutput: input.testOutput,
+      testCommand: input.testCommand,
       diagnosisReasoning: input.diagnosisReasoning,
       priorIterationsBlock: input.priorIterationsBlock,
       acceptanceTestPath: input.acceptanceTestPath,
-      testFileContent: input.testFileContent,
     });
     return {
       role: { id: "role", content: "", overridable: false },
@@ -61,11 +61,11 @@ export const acceptanceFixTestOp: RunOperation<AcceptanceFixTestInput, Acceptanc
   build(input, _ctx) {
     const prompt = new AcceptancePromptBuilder().buildTestFixPrompt({
       testOutput: input.testOutput,
+      testCommand: input.testCommand,
       diagnosisReasoning: input.diagnosisReasoning,
       priorIterationsBlock: input.priorIterationsBlock,
       failedACs: input.failedACs,
       acceptanceTestPath: input.acceptanceTestPath,
-      testFileContent: input.testFileContent ?? "",
     });
     return {
       role: { id: "role", content: "", overridable: false },

--- a/src/pipeline/stages/acceptance.ts
+++ b/src/pipeline/stages/acceptance.ts
@@ -264,13 +264,9 @@ export const acceptanceStage: PipelineStage = {
             abortSignal: ctx.abortSignal,
           });
           hardeningRetries = result.promoted.length;
-          logger.info("acceptance", "Hardening pass complete", {
-            storyId: ctx.story.id,
-            promoted: result.promoted.length,
-            discarded: result.discarded.length,
-          });
         } catch (err) {
-          logger.warn("acceptance", "Hardening pass failed (non-blocking)", {
+          // runHardeningPass already logs "Hardening pass failed" with full storyIds attribution
+          logger.debug("acceptance", "Hardening pass failed (non-blocking)", {
             storyId: ctx.story.id,
             error: err instanceof Error ? err.message : String(err),
           });

--- a/src/prompts/builders/acceptance-builder-helpers.ts
+++ b/src/prompts/builders/acceptance-builder-helpers.ts
@@ -1,0 +1,64 @@
+import { detectFramework, formatFailureSummary, parseTestOutput } from "../../test-runners";
+
+const MAX_FAILURE_CHARS = 4000;
+const TAIL_FALLBACK_LINES = 60;
+const MAX_ENV_FAILURE_CHARS = 4000;
+
+/**
+ * Convert raw test-runner stdout into a compact, language-agnostic failure summary.
+ *
+ * Decision tree:
+ *   1. structured failures available → formatFailureSummary
+ *   2. failed > 0 but no structured failures → tail of last N lines (failures cluster at end)
+ *   3. no failures AND no passes → likely environmental failure (compile error, missing binary)
+ *   4. all passed → summary header only (caller invoked fix path unexpectedly)
+ */
+export function formatTestOutputForFix(rawOutput: string): string {
+  const summary = parseTestOutput(rawOutput);
+  const framework = detectFramework(rawOutput);
+  const header = `Test runner: ${framework}\nResult: ${summary.passed} passed, ${summary.failed} failed`;
+
+  if (summary.failures.length > 0) {
+    return `${header}\n\nFailures:\n${formatFailureSummary(summary.failures, MAX_FAILURE_CHARS)}`;
+  }
+
+  if (summary.failed > 0) {
+    const lines = rawOutput.trim().split("\n");
+    const tail = lines.slice(-TAIL_FALLBACK_LINES).join("\n");
+    return `${header}\n\nTest output (last ${TAIL_FALLBACK_LINES} lines — structured parse unavailable):\n${tail}`;
+  }
+
+  if (summary.passed === 0) {
+    const capped =
+      rawOutput.length > MAX_ENV_FAILURE_CHARS
+        ? `${rawOutput.slice(0, MAX_ENV_FAILURE_CHARS)}\n... (truncated — environmental failure suspected)`
+        : rawOutput;
+    return `${header}\n\nNo structured tests detected — environmental failure suspected:\n${capped}`;
+  }
+
+  return header;
+}
+
+const LANG_BY_EXT: Record<string, string> = {
+  ".ts": "typescript",
+  ".tsx": "typescript",
+  ".js": "javascript",
+  ".jsx": "javascript",
+  ".mjs": "javascript",
+  ".cjs": "javascript",
+  ".go": "go",
+  ".py": "python",
+  ".rs": "rust",
+  ".rb": "ruby",
+  ".java": "java",
+  ".kt": "kotlin",
+  ".swift": "swift",
+  ".php": "php",
+};
+
+/** Derive the markdown code-fence language hint from a file path extension. Returns "" for unknown extensions. */
+export function fenceLangFor(filePath: string | undefined): string {
+  if (!filePath) return "";
+  const ext = filePath.match(/\.[^./]+$/)?.[0] ?? "";
+  return LANG_BY_EXT[ext.toLowerCase()] ?? "";
+}

--- a/src/prompts/builders/acceptance-builder.ts
+++ b/src/prompts/builders/acceptance-builder.ts
@@ -11,7 +11,9 @@
  */
 
 import type { PRD } from "../../prd/types";
+import { buildTestFrameworkHint } from "../../test-runners";
 import { wrapJsonPrompt } from "../../utils/llm-json";
+import { formatTestOutputForFix } from "./acceptance-builder-helpers";
 
 export type AcceptanceRole = "generator" | "diagnoser" | "fix-executor";
 
@@ -65,6 +67,7 @@ export interface FixGeneratorParams {
 export interface DiagnosisPromptParams {
   testOutput: string;
   testFileContent: string;
+  acceptanceTestPath?: string;
   sourceFiles: Array<{ path: string; content: string }>;
   /** Minimal shape — avoids importing SemanticVerdict across layers */
   semanticVerdicts?: Array<{ storyId: string; passed: boolean }>;
@@ -96,7 +99,7 @@ export interface GeneratorFromSpecParams {
 
 export interface DiagnosisTemplateParams {
   truncatedOutput: string;
-  testFileContent: string;
+  acceptanceTestPath: string;
   sourceFilesSection: string;
   verdictSection: string;
   maxFileLines: number;
@@ -104,19 +107,19 @@ export interface DiagnosisTemplateParams {
 
 export interface SourceFixParams {
   testOutput: string;
+  testCommand?: string;
   diagnosisReasoning?: string;
   priorIterationsBlock?: string;
   acceptanceTestPath: string;
-  testFileContent?: string;
 }
 
 export interface TestFixParams {
   testOutput: string;
+  testCommand?: string;
   diagnosisReasoning?: string;
   priorIterationsBlock?: string;
   failedACs: string[];
   acceptanceTestPath: string;
-  testFileContent: string;
 }
 
 // ─── Builder ──────────────────────────────────────────────────────────────────
@@ -187,10 +190,9 @@ TASK: Diagnose whether the failure is due to a bug in the SOURCE CODE or a bug i
 FAILING TEST OUTPUT:
 ${p.truncatedOutput}
 
-ACCEPTANCE TEST FILE CONTENT:
-\`\`\`typescript
-${p.testFileContent}
-\`\`\`
+ACCEPTANCE TEST FILE: ${p.acceptanceTestPath}
+
+(Use Read on the path above to inspect the test code if needed for diagnosis.)
 
 SOURCE FILES (auto-detected from imports, up to ${p.maxFileLines} lines each):
 ${p.sourceFilesSection}
@@ -201,14 +203,14 @@ ${responseSchema}`;
 
   /** Prompt for acceptanceFixSourceOp — instructs agent to fix source implementation. */
   buildSourceFixPrompt(p: SourceFixParams): string {
-    let prompt = `ACCEPTANCE TEST FAILURE:\n${p.testOutput}\n\n`;
+    let prompt = "ACCEPTANCE TEST FAILURE — fix the source implementation.\n\n";
+    if (p.testCommand) prompt += `Test framework: ${buildTestFrameworkHint(p.testCommand)}\n\n`;
+    prompt += `TEST OUTPUT:\n${formatTestOutputForFix(p.testOutput)}\n\n`;
     if (p.diagnosisReasoning) prompt += `DIAGNOSIS:\n${p.diagnosisReasoning}\n\n`;
     if (p.priorIterationsBlock) prompt += p.priorIterationsBlock;
     prompt += `ACCEPTANCE TEST FILE: ${p.acceptanceTestPath}\n\n`;
-    if (p.testFileContent && p.testFileContent.length > 0) {
-      prompt += `\`\`\`typescript\n${p.testFileContent}\n\`\`\`\n\n`;
-    }
-    prompt += "Fix the source implementation. Do NOT modify the test file.";
+    prompt += "Read the test file at the path above for context, then fix the source implementation. ";
+    prompt += "Do NOT modify the test file.";
     return prompt;
   }
 
@@ -274,7 +276,7 @@ Respond with ONLY the fix description (no JSON, no markdown, just the descriptio
 
     return this.buildDiagnosisPromptTemplate({
       truncatedOutput,
-      testFileContent: p.testFileContent,
+      acceptanceTestPath: p.acceptanceTestPath ?? "(path unavailable — inspect test output for file references)",
       sourceFilesSection,
       verdictSection,
       maxFileLines: MAX_FILE_LINES,
@@ -366,12 +368,13 @@ Assert about HTTP responses, status codes, and API endpoint output.${framework}
   buildTestFixPrompt(p: TestFixParams): string {
     let prompt = "ACCEPTANCE TEST BUG — surgical fix required.\n\n";
     prompt += `FAILING ACS: ${p.failedACs.join(", ")}\n\n`;
-    prompt += `TEST OUTPUT:\n${p.testOutput}\n\n`;
+    if (p.testCommand) prompt += `Test framework: ${buildTestFrameworkHint(p.testCommand)}\n\n`;
+    prompt += `TEST OUTPUT:\n${formatTestOutputForFix(p.testOutput)}\n\n`;
     if (p.diagnosisReasoning) prompt += `DIAGNOSIS:\n${p.diagnosisReasoning}\n\n`;
     if (p.priorIterationsBlock) prompt += p.priorIterationsBlock;
     prompt += `ACCEPTANCE TEST FILE: ${p.acceptanceTestPath}\n\n`;
-    prompt += `\`\`\`typescript\n${p.testFileContent}\n\`\`\`\n\n`;
-    prompt += "Fix ONLY the failing test assertions for the ACs listed above. ";
+    prompt += "Read the test file at the path above before editing. The fix should be ";
+    prompt += "surgical — locate the failing AC blocks and adjust their assertions only. ";
     prompt += "Do NOT modify passing tests. Do NOT modify source code. ";
     prompt += "Edit the test file in place.";
     return prompt;

--- a/src/prompts/builders/acceptance-builder.ts
+++ b/src/prompts/builders/acceptance-builder.ts
@@ -66,7 +66,12 @@ export interface FixGeneratorParams {
 
 export interface DiagnosisPromptParams {
   testOutput: string;
-  testFileContent: string;
+  /**
+   * @deprecated No longer embedded in the diagnosis prompt — replaced by path-only reference.
+   * Retained at the call layer only for import-parsing in loadSourceFilesForDiagnosis.
+   * Will be removed when that utility migrates to path-based parsing.
+   */
+  testFileContent?: string;
   acceptanceTestPath?: string;
   sourceFiles: Array<{ path: string; content: string }>;
   /** Minimal shape — avoids importing SemanticVerdict across layers */

--- a/src/prompts/index.ts
+++ b/src/prompts/index.ts
@@ -33,6 +33,7 @@ export type {
   DiagnosisPromptParams,
   RefinementPromptOptions,
 } from "./builders/acceptance-builder";
+export { fenceLangFor, formatTestOutputForFix } from "./builders/acceptance-builder-helpers";
 
 // Rectifier prompt builder — cross-domain rectification for TDD, verify, and review triggers.
 export { RectifierPromptBuilder, CONTRADICTION_ESCAPE_HATCH } from "./builders/rectifier-builder";

--- a/src/prompts/index.ts
+++ b/src/prompts/index.ts
@@ -33,6 +33,7 @@ export type {
   DiagnosisPromptParams,
   RefinementPromptOptions,
 } from "./builders/acceptance-builder";
+// fenceLangFor: no production call site yet — prepared for follow-up F3 (SSOT fence-lang across all prompt builders).
 export { fenceLangFor, formatTestOutputForFix } from "./builders/acceptance-builder-helpers";
 
 // Rectifier prompt builder — cross-domain rectification for TDD, verify, and review triggers.

--- a/test/unit/execution/lifecycle/acceptance-loop-cycle.test.ts
+++ b/test/unit/execution/lifecycle/acceptance-loop-cycle.test.ts
@@ -451,15 +451,15 @@ describe("strategy buildInput closures", () => {
       makePrd(),
       { failedACs: ["AC-1"], testOutput: "initial output" },
       makeDiagnosis(),
-      "test-content",
       "/path/to/test.ts",
+      "bun test",
     );
 
     const sourceStrategy = capturedCycle!.strategies[0];
     const input = sourceStrategy.buildInput([], [], {} as never) as Record<string, unknown>;
     expect(input.testOutput).toBe("initial output");
     expect(input.acceptanceTestPath).toBe("/path/to/test.ts");
-    expect(input.testFileContent).toBe("test-content");
+    expect(input.testCommand).toBe("bun test");
   });
 
   test("source-fix buildInput includes prior iterations block", async () => {
@@ -474,7 +474,6 @@ describe("strategy buildInput closures", () => {
       makePrd(),
       { failedACs: ["AC-1"], testOutput: "initial output" },
       makeDiagnosis(),
-      "test-content",
       "/path/to/test.ts",
     );
 

--- a/test/unit/prompts/acceptance-builder.test.ts
+++ b/test/unit/prompts/acceptance-builder.test.ts
@@ -150,7 +150,7 @@ describe("builder.buildGeneratorFromSpecPrompt()", () => {
 describe("builder.buildDiagnosisPromptTemplate()", () => {
   const base = {
     truncatedOutput: "FAIL: AC-1 assertion error",
-    testFileContent: 'import { test } from "bun:test"; test("AC-1: x", () => {});',
+    acceptanceTestPath: "/project/.nax/features/feat/.nax-acceptance.test.ts",
     sourceFilesSection: "(No source files could be resolved from imports)",
     verdictSection: "",
     maxFileLines: 500,
@@ -177,10 +177,15 @@ describe("builder.buildDiagnosisPromptTemplate()", () => {
       expect(result).toContain(base.truncatedOutput);
     });
 
-    test("includes test file content in fenced block", () => {
+    test("references acceptance test path (Bug 6 — no embedded body)", () => {
       const result = builder.buildDiagnosisPromptTemplate(base);
-      expect(result).toContain("```typescript");
-      expect(result).toContain(base.testFileContent);
+      expect(result).toContain(base.acceptanceTestPath);
+      expect(result).not.toContain("```typescript");
+    });
+
+    test("instructs agent to use Read on the test path", () => {
+      const result = builder.buildDiagnosisPromptTemplate(base);
+      expect(result).toContain("Read");
     });
 
     test("includes source files section", () => {
@@ -219,14 +224,29 @@ describe("builder.buildDiagnosisPromptTemplate()", () => {
 
 describe("builder.buildSourceFixPrompt()", () => {
   const base = {
-    testOutput: "FAIL: AC-1 null pointer",
+    testOutput: "(fail) AC-1: null pointer [2ms]\n  Error: Cannot read property\n\n 0 pass\n 1 fail",
     diagnosisReasoning: "Source file has uninitialized field",
     acceptanceTestPath: "/project/.nax/features/feat/.nax-acceptance.test.ts",
-    testFileContent: 'test("AC-1: x", () => { expect(foo()).toBe(1); });',
   };
 
-  test("includes test output", () => {
-    expect(builder.buildSourceFixPrompt(base)).toContain(base.testOutput);
+  test("includes structured test output (Bug 6 regression)", () => {
+    const result = builder.buildSourceFixPrompt(base);
+    expect(result).toContain("AC-1");
+    expect(result).toContain("Cannot read property");
+  });
+
+  test("does not embed test file content (Bug 6 regression)", () => {
+    const result = builder.buildSourceFixPrompt(base);
+    expect(result).not.toContain("```typescript");
+  });
+
+  test("references acceptance test path", () => {
+    expect(builder.buildSourceFixPrompt(base)).toContain(base.acceptanceTestPath);
+  });
+
+  test("instructs agent to Read the test file", () => {
+    const result = builder.buildSourceFixPrompt(base);
+    expect(result).toContain("Read the test file at the path above");
   });
 
   test("includes diagnosis reasoning", () => {
@@ -239,19 +259,9 @@ describe("builder.buildSourceFixPrompt()", () => {
     expect(result).toContain("prior table");
   });
 
-  test("includes acceptance test path", () => {
-    expect(builder.buildSourceFixPrompt(base)).toContain(base.acceptanceTestPath);
-  });
-
-  test("includes test file content in fenced typescript block", () => {
-    const result = builder.buildSourceFixPrompt(base);
-    expect(result).toContain("```typescript");
-    expect(result).toContain(base.testFileContent);
-  });
-
-  test("omits fenced block when testFileContent is empty", () => {
-    const result = builder.buildSourceFixPrompt({ ...base, testFileContent: "" });
-    expect(result).not.toContain("```typescript");
+  test("includes test framework hint when testCommand is provided", () => {
+    const result = builder.buildSourceFixPrompt({ ...base, testCommand: "bun test" });
+    expect(result).toContain("Test framework:");
   });
 
   test("instructs not to modify test file", () => {
@@ -263,22 +273,36 @@ describe("builder.buildSourceFixPrompt()", () => {
 
 describe("builder.buildTestFixPrompt()", () => {
   const base = {
-    testOutput: "FAIL: AC-1 assertion error",
+    testOutput: "(pass) AC-1: ok [1ms]\n(fail) AC-2: assertion failed [2ms]\n  Error: Expected 1 got 0\n\n 1 pass\n 1 fail",
     diagnosisReasoning: "Test uses wrong assertion type",
-    failedACs: ["AC-1", "AC-3"],
+    failedACs: ["AC-2"],
     acceptanceTestPath: "/project/.nax/features/feat/.nax-acceptance.test.ts",
-    testFileContent: 'test("AC-1: x", () => { expect(foo()).toBe(1); });',
   };
 
   test("includes failing ACs", () => {
     const result = builder.buildTestFixPrompt(base);
-    expect(result).toContain("AC-1");
-    expect(result).toContain("AC-3");
+    expect(result).toContain("AC-2");
   });
 
-  test("includes test output", () => {
+  test("does not embed test file content (Bug 6 regression)", () => {
     const result = builder.buildTestFixPrompt(base);
-    expect(result).toContain(base.testOutput);
+    expect(result).not.toContain("```typescript");
+  });
+
+  test("drops (pass) lines from test output (Bug 6 regression)", () => {
+    const result = builder.buildTestFixPrompt(base);
+    expect(result).not.toContain("(pass) AC-1");
+    expect(result).toContain("AC-2");
+    expect(result).toContain("Expected 1 got 0");
+  });
+
+  test("references acceptance test path", () => {
+    expect(builder.buildTestFixPrompt(base)).toContain(base.acceptanceTestPath);
+  });
+
+  test("instructs agent to Read the test file", () => {
+    const result = builder.buildTestFixPrompt(base);
+    expect(result).toContain("Read the test file at the path above");
   });
 
   test("includes diagnosis reasoning", () => {
@@ -292,15 +316,14 @@ describe("builder.buildTestFixPrompt()", () => {
     expect(result).toContain("prior table");
   });
 
-  test("includes test file content in fenced typescript block", () => {
-    const result = builder.buildTestFixPrompt(base);
-    expect(result).toContain("```typescript");
-    expect(result).toContain(base.testFileContent);
+  test("includes test framework hint when testCommand is provided", () => {
+    const result = builder.buildTestFixPrompt({ ...base, testCommand: "bun test" });
+    expect(result).toContain("Test framework:");
   });
 
   test("instructs to fix only failing ACs and not source code", () => {
     const result = builder.buildTestFixPrompt(base);
-    expect(result).toContain("Fix ONLY the failing test assertions");
+    expect(result).toContain("surgical");
     expect(result).toContain("Do NOT modify source code");
   });
 });

--- a/test/unit/prompts/builders/acceptance-builder-helpers.test.ts
+++ b/test/unit/prompts/builders/acceptance-builder-helpers.test.ts
@@ -1,0 +1,82 @@
+import { describe, expect, test } from "bun:test";
+import { fenceLangFor, formatTestOutputForFix } from "../../../../src/prompts";
+
+// ─── formatTestOutputForFix ───────────────────────────────────────────────────
+
+describe("formatTestOutputForFix", () => {
+  test("bun test output: structured failures extracted, (pass) lines excluded", () => {
+    const raw = [
+      "(pass) AC-1: should return empty array [1ms]",
+      "(fail) AC-2: should handle edge case [2ms]",
+      "  Error: Expected 0 but got 1",
+      "",
+      " 1 pass",
+      " 1 fail",
+    ].join("\n");
+
+    const out = formatTestOutputForFix(raw);
+    expect(out).toContain("AC-2");
+    expect(out).toContain("Expected 0 but got 1");
+    expect(out).not.toContain("(pass) AC-1");
+  });
+
+  test("result is significantly smaller than raw output with passing tests", () => {
+    const passLines = Array.from({ length: 36 }, (_, i) => `(pass) AC-${i + 1}: test [1ms]`).join("\n");
+    const failLine = "(fail) AC-37: broken test [2ms]\n  Error: nope";
+    const raw = `${passLines}\n${failLine}\n\n 36 pass\n 1 fail`;
+
+    const out = formatTestOutputForFix(raw);
+    expect(out.length).toBeLessThan(raw.length);
+    expect(out).toContain("AC-37");
+    expect(out).not.toContain("(pass) AC-1");
+  });
+
+  test("unknown framework with failures falls back to tail lines", () => {
+    const lines = Array.from({ length: 100 }, (_, i) => `line ${i}`);
+    lines.push("FAILED: some test");
+    lines.push("some_test.go:42: assertion failed");
+    const raw = lines.join("\n");
+
+    const out = formatTestOutputForFix(raw);
+    expect(out).toContain("some_test.go:42");
+    expect(out).not.toContain("line 0");
+  });
+
+  test("environmental failure (compile error, no tests ran)", () => {
+    const raw = "error: cannot find module 'foo'\nat line 5\nBuild failed";
+    const out = formatTestOutputForFix(raw);
+    expect(out).toContain("environmental failure suspected");
+  });
+
+  test("output is bounded for very large raw input", () => {
+    const raw = "x".repeat(500_000);
+    const out = formatTestOutputForFix(raw);
+    expect(out.length).toBeLessThan(10_000);
+  });
+
+  test("all tests passed (unexpected call to fix path) returns only header", () => {
+    const raw = "(pass) AC-1: ok [1ms]\n(pass) AC-2: ok [1ms]\n\n 2 pass\n 0 fail";
+    const out = formatTestOutputForFix(raw);
+    expect(out).toContain("2 passed");
+    expect(out).toContain("0 failed");
+    expect(out).not.toContain("Failures:");
+  });
+});
+
+// ─── fenceLangFor ─────────────────────────────────────────────────────────────
+
+describe("fenceLangFor", () => {
+  test.each([
+    [".nax-acceptance.test.ts", "typescript"],
+    ["foo_test.go", "go"],
+    ["test_foo.py", "python"],
+    ["some_spec.rs", "rust"],
+    ["Module.java", "java"],
+    ["Main.kt", "kotlin"],
+    ["app.rb", "ruby"],
+    ["weird.unknown", ""],
+    [undefined, ""],
+  ])("fenceLangFor(%s) → %s", (input, expected) => {
+    expect(fenceLangFor(input as string | undefined)).toBe(expected);
+  });
+});

